server-nonces: first draft (#1)
pmhsfelix/draft-ietf-oauth-attestation-based-client-auth: draft-ietf-oauth-attestation-based-client-auth.md

4. The JWT MUST contain an "aud" (audience) claim containing a value that identifies the authorization server as an intended audience. The {{RFC8414}} issuer identifier URL of the authorization server MUST be used as a value for an "aud" element to identify the authorization server as the intended audience of the JWT.

> 5. The JWT MAY contain an "nonce" claim containing a string value that is provided by the authorization server to associate the Client Attestation PoP JWT with a particular transaction and prevent replay attacks.

Reviewer: ```suggestion
5. The JWT MAY contain a "nonce" claim with a string value that is provided by the authorization server to associate the Client Attestation PoP JWT with a particular transaction and prevent replay attacks.
```

}
~~~

# Authorization Server-provided nonces

> This section specifies a mechanism using server provided nonces to limit the lifetime of a Client Attestation PoP JWT, similar to the one defined in section 8 of {{RFC9449}}.

Reviewer: RFC9449 specifically states that the nonce value is opaque. I think this spec should also define the nonce as an opaque value.

> This section specifies a mechanism using server provided nonces to limit the lifetime of a Client Attestation PoP JWT, similar to the one defined in section 8 of {{RFC9449}}.

Reviewer: Maybe add something along this line here:
A nonce value is a string as defined in (reference to nonce-syntax) that is opaque to the client.

Without employing such a mechanism, a malicious party controlling the client (potentially including the end-user) can create Client Attestation PoP JWT for use arbitrarily far in the future.
In addition, a mismatch between the time reference of a non-malicious client (e.g. a mobile device) and the time reference of the authorization server, may result in Client Attestation PoP JWT that are never accepted by the authorization server.

Including a nonce value contributed by the authorization server in the Client Attestation PoP JWT MAY be used by authorization servers to limit the lifetime of those proofs, both protecting against attacks and allowing for interoperability with clients with different time references.

> Nonces are included in the `nonce` JWT payload claim, as JSON strings.

Reviewer: I'd remove this line from here. The value type of the `nonce` claim is describe above. Instead provide instructions for the client in the next section on how to process the server's response.

Reviewer: My main concern with this line is that it uses plural. The `nonce` claim must only contain one nonce value. The current wording may imply that several nonces (nonce values) are included in a nonce claim...

An authorization server MAY require the use of nonces.
This requirement is communicated to the client by having the authorization server return an HTTP response with the `400` status and an `error` field with the value `"use_attestation_nonce"`, when the Client Attestation PoP JWT present in the HTTP request does not contain a valid nonce.
This HTTP response MUST also contain the `Attestation-Nonce` header, with the value of the new server provided nonce.

> The client MUST use this new nonce value to create a new Client Attestation PoP JWT and resend the previous request.

Reviewer: ```suggestion
The client MUST use this new nonce value to create a new Client Attestation PoP JWT containing the `nonce` claim with the supplied nonce value. The client MUST resend the previous request using this new Client Attestation PoP JWT.
```

> The following non-normative example presents an example of a non-success response, including the new nonce in the `Attestation-Nonce` header.

Reviewer: ```suggestion
The following non-normative example presents a non-success response that contains the new nonce value in the `Attestation-Nonce` header.
```

~~~
HTTP/1.1 400 Bad Request
Attestation-Nonce: eyJ7S_zG.eyJH0-Z.HX4w-7v

{
  "error": "use_attestation_nonce",
  "error_description":
    "Authorization server requires nonce in Client Attestation PoP JWT proof"
}
~~~

> Authorization servers MAY also include the `Attestation-Nonce` in success responses, as a way to communicate newer nonces values to clients.

Reviewer: How are clients supposed to act upon a success response that includes an `Attestation-Nonce` header? In an error-response they must update the Client Attestation PoP JWT and retry the request to fix the error. In this success case, there is no error to fix. What should the client do with the `Attestation-Nonce` header?

Author: Use the received nonces in future requests. I'll make that clearer.

Clients SHOULD keep using a nonce received in a `Attestation-Nonce` header until a newer value is received via this header, both on success or non-success responses.

> Responses that include the `Attestation-Nonce` HTTP header should be uncacheable (e.g., using `Cache-Control: no-store` in response to a GET request) to prevent the response from being used to serve a subsequent request and a stale nonce value from being used as a result.

Reviewer: Is the uncachable requirement normative?

Author: Following what is in the DPoP spec.
